### DISCUSSION TITLE
Improve the explorability of content within the app

### DISCUSSION THREAD

**Discussion by Lurux** [z-huang/music#871] [Ideas]
@z-huang - continuation of the discussion on #869

---

> I think 6 items is enough to have a rough view of your taste. Viewing all is not necessary.

I disagree. I personally view the stats screen, not just as a quick way to view my tastes, but as a way to explore the content that I listen to on the app. 6 items is not enough for that.

As an example, I previously relied upon the fact that the artists list was generated automatically from the songs in the library. Now that is gone, I need to bookmark them manually, but there isn't anywhere in the app to browse the artists I might want to save ! The stats screen seems like the best place for that, but 6 items simply isn't enough.

I personally think having an unlimited (or eg. 100 items) stats screen for each item of the app, combined with a quick access from the library and the ability to bookmark or remove items directly from these screens, would be a terrific improvement when browsing for content to bookmark in the app.

Some users may also want to keep a limited number of bookmarked artists/albums, but still have access to a top list of items generated from the history - similar to the old behavior of the artists tab with "play time" sorting. This proposal also addresses this use-case.

If you have different plans to address these issues, or have any concerns about the UX or maintainability of this, we can talk about it !

---

> We can only get the id and name of the most played albums, so we have to make network requests to get the full information of the albums, which takes more time.

When exploring the codebase I noticed artists and albums are treated differently: it seems like artists from the history are systematically saved to the DB, then the bookmarkedAt field is used to denote the bookmarked status of the artist. Meanwhile, it seems like the albums are only added to the DB when saved to the library.

I don't know how difficult that would be, but maybe we should bring albums in line with the artists ? This would solve the stats screen loading problem, and might make the project more maintainable in the long run, as having two similar things work completely differently sound like it will quickly get confusing.

**Comment by z-huang** [comment]
Sorry for the late reply because I was considering it thoroughly. Actually, the stats screen is inspired by Pano Scrobbler's charts, and it has buttons to view all items. Therefore, I accept that InnerTune should have the buttons as well. It can be one way to explore content.

> Once that is done, I'd suggest adding links to these pages directly from the main tabs of the library in order to improve the navigation and discoverability of content within the app

I'd like to know where you would add these links, thanks.

> As an example, I previously relied upon the fact that the artists list was generated automatically from the songs in the library. Now that is gone, I need to bookmark them manually, but there isn't anywhere in the app to browse the artists I might want to save !

True, it's a bit troublesome to add artists one by one. Therefore, I added filters in library artists screen. (412c883)

> I don't know how difficult that would be, but maybe we should bring albums in line with the artists ? This would solve the stats screen loading problem, and might make the project more maintainable in the long run, as having two similar things work completely differently sound like it will quickly get confusing.

Yes, in order to fix the stats screen loading problem and implement some new features, the app should save albums in database just like artists.

**Comment by Lurux** [comment]
> Sorry for the late reply

No problem ! I wanted to experiment with this on my side, but I haven't had the time to do much because I've been kinda busy lately.

---

Regarding my "full stats screen page" proposal:

> It has buttons to view all items. Therefore, I accept that InnerTune should have the buttons as well

Well, the new filter system pretty much addresses the use case I had for this, so I don't see a pressing need to implement this anymore. The main reason I suggested to have a link to stats screen pages rather than an "all library" filter is because this seemed to be the direction you were going for, but I actually think the filter system is more usable for this particular use case.

Stats screen pages would still have the benefit of showing a specific period and not being limited to the items in the library, so we could implement these down the road. However, I no longer think is necessary for the usability of the app, especially if filters can be added to the other tabs (see below).

---

Regarding the albums tab:

> The app should save albums in database just like artists

Since we were talking about bringing albums in line with the artists, I think it would be really good to bring the same filters to the albums tab. I suggest adding the filters to this tab even if we can't bring the backend in line with artists for the next release, and limit the "all" filter to a certain number of items, so it doesn't take ages to load but at least the functionality is there.

---

Bringing everything together:

Still rolling with the idea of filters in the library tabs, I'd like to suggest additional changes to bring more consistency in the app:

- Add "library / liked" filters to the songs tab
- Rename the current "all" filter to "library" (this seems clearer to me)
- Change "bookmark artist" to "like artist" (consistency)
- Ultimately, change "add album to library" to "like album" (consistency)
- Hypothetically we could add a "downloaded" filter to the songs tab, too

The idea is that, for each library tab, the "library" filter would show items generated from the list of songs in the library, while the "liked" filter would show the liked items in this category.

Basically, here is how this could relate to the current state of the app:

- Songs tab
  - Library filter (current songs tab)
  - Liked filter (current playlists > liked)
- Artists tab
  - Library filter (current artists > all)
  - Liked filter (current artists > bookmarked)
- Albums tab
  - Library filter (doesn't exist yet)
  - Liked filter (current albums tab)

Basically this would bring more consistency to the app, and we could add more filters in the future if we want to (eg. history for all items in the history, downloaded, explore for youtube suggestions, etc).

**Reply by z-huang** [reply]
Done!

**Comment by Lurux** [comment]
As a side note: it would be really cool if we could browse artist pages when offline ! This is the one thing that works better with albums currently.

Since the basic details are already saved anyways, we could just show the header and "from your library" sections, even if the other sections obviously can't be loaded.

I don't know how to do that, but I hope this can be implemented without too much hassle. It would tremendously improve the offline functionality of the app !
